"Id hash mismatch" if helios-client/helios-testing uses Jackson 2.9.x while helios-master/helios-solo uses Jackson 2.8.x

The logic in Job.Builder that creates the "hash" component of the JobId [does the following](https://github.com/spotify/helios/blob/29106a7e5f290a37152ef0e28b21758ed5d0848f/helios-client/src/main/java/com/spotify/helios/common/Json.java#L344-L352):

- serialize the non-metadata parameters of the Job into a JSON string
- deserialize this string back into a Map<String, Object>
- serialize this Map into JSON bytes using `objectWriter.writeValueAsBytes()`. The objectWriter instance has been configured with `setSerializationInclusion(JsonInclude.Include.NON_EMPTY)`.

When the Create Job API method is called, the helios-master re-calculates the hash for the submitted Job on the server-side and verifies that it is equal to the hash submitted by the client. If the two values do not match, a CreateJobResponse is sent back with a status of `INVALID_JOB_DEFINITION` and an errors list containing `Id hash mismatch: hash1 != hash2`.

The same logic is performed by helios-master when run standalone and also when run inside of helios-solo (as a docker image) in integration tests via helios-testing.

A Job or TemporaryJob (with helios-testing) with a `registration` section like

```java
temporaryJobs.job()
  .port("http", 8080)
  .registration("foobar", "tcp", "http")
```

is represented in JSON like `"registration": {"foobar/tcp":{"ports":{"http":{}}}}`.

With Jackson 2.8.x, which is currently used in helios (-client, -master, -testing, -solo), this portion of the JSON is preserved in each step of the JobId hash calculation mentioned below.

With Jackson 2.9.x, the last step of the JobID hash calculation (where the Map is serialized into JSON bytes) omits the entire `"registration"` field, likely because it determines that the inner object of `{"http":{}}` is empty, which then (theoretically) propagates outward. See https://github.com/FasterXML/jackson-databind/issues/1444 and https://github.com/FasterXML/jackson-databind/issues/1736.

Since the JobID hash calculation relies on Jackson when calculating the hash, and creating a Job requires the client and the server to calcualte the same JobID hash, there is a need for both the client and the server to both be using a Jackson version <= 2.8.x or both be using >= 2.9.x. The same applies for tests that use helios-testing/TemporaryJobs (the client) and helios-solo (the server).

If a helios-testing user upgrades their Maven project to use Jackson 2.9.x, then their tests that attempt to deploy TemporaryJobs will fail with 

> CreateJobResponse{status=INVALID_JOB_DEFINITION, errors=[Id hash mismatch: ... ]

because the helios-master inside the helios-solo docker image is using Jackson 2.8.8.

If we upgrade helios to use Jackson 2.9.x, then any users with existing helios-testing tests where the project uses Jackson version <= 2.8.x will download `spotify/helios-solo:latest` the next time they run their tests and begin to get `CreateJobResponse{status=INVALID_JOB_DEFINITION...}` since the `latest` docker image is pulled on each run. Additionally, when a helios-master is deployed using Jackson 2.9.x, any existing helios-cli users of previous versions (which Jackson 2.8.x bundled inside) will be submitting Jobs with an ID hash that the server calculates a different value for.

Arguably the Jackson behavior in 2.8.x with this type of JSON and `setSerializationInclusion(JsonInclude.Include.NON_EMPTY)` (changed in https://github.com/FasterXML/jackson-databind/issues/1444) was incorrect all along, but Helios's insistence that the client and the server calculate the same hash (while using third party libraries which could change serialization behavior with any version upgrade) has created a situation where we need a helios-testing user to be aware and locked into the Jackson version used in the master branch of Helios, and where we need helios-cli/helios-client users to be using a version of Jackson that behaves the same way as the version used inside helios-master.

I'm not fully sure yet but https://github.com/FasterXML/jackson-databind/issues/1444 makes it sound like it might be possible to use `setDefaultPropertyInclusion()` (introduced in 2.9.x I think?) to get the same serialization behavior with these empty objects as we had in 2.8.x.

The larger issue here is that the JobId hash - which is of dubious value - relies on a third-party library for which a different version is frequently used in projects that use helios-testing.

Upgrading helios to use 2.9.0 causes the JobIdHashIsStableTest to fail as expected (that trip-wire is working):

```diff
$> git diff
diff --git a/pom.xml b/pom.xml
index 5aa68241..23d41d43 100644
--- a/pom.xml
+++ b/pom.xml
@@ -157,32 +157,32 @@
             <dependency>
                 <groupId>com.fasterxml.jackson.dataformat</groupId>
                 <artifactId>jackson-dataformat-yaml</artifactId>
-                <version>2.8.8</version>
+                <version>2.9.0</version>
             </dependency>
             <dependency>
                 <groupId>com.fasterxml.jackson.datatype</groupId>
                 <artifactId>jackson-datatype-guava</artifactId>
-                <version>2.8.8</version>
+                <version>2.9.0</version>
             </dependency>
             <dependency>
                 <groupId>com.fasterxml.jackson.core</groupId>
                 <artifactId>jackson-core</artifactId>
-                <version>2.8.8</version>
+                <version>2.9.0</version>
             </dependency>
             <dependency>
                 <groupId>com.fasterxml.jackson.core</groupId>
                 <artifactId>jackson-annotations</artifactId>
-                <version>2.8.8</version>
+                <version>2.9.0</version>
             </dependency>
             <dependency>
                 <groupId>com.fasterxml.jackson.core</groupId>
                 <artifactId>jackson-databind</artifactId>
-                <version>2.8.8</version>
+                <version>2.9.0</version>
             </dependency>
             <dependency>
                 <groupId>com.fasterxml.jackson.jaxrs</groupId>
                 <artifactId>jackson-jaxrs-json-provider</artifactId>
-                <version>2.8.8</version>
+                <version>2.9.0</version>
             </dependency>
             <dependency>
                 <groupId>joda-time</groupId>
```

```
$> mvn -am -pl helios-client test
...

Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.811 sec - in com.spotify.helios.common.ResolverTest

Results :

Failed tests:
  JobIdHashIsStableTest.testInitialFullJob:94->assertExpectedHash:84 If this hash changed, you are probably introducing a backwards-incompatible change to the Job class! expected:<[9693b4579ac48ed41f317ecf5edc5496b005b79d]> but was:<[117a46e13ac259b5fcceac23722d6987e8ad44f7]>
  JobValidatorTest.testJobFromJsonWithInvalidRegistration:535
Expected: iterable containing ["registration for 'volumes' is malformed: does not have a port mapping"]
     but: Not matched: "Id hash mismatch: 4ee3a68429951bb9603b753c73fcefab27420dbe != 196821077786559b63f0f65fc511105fded4e347"

Tests run: 114, Failures: 2, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] Helios Parent ...................................... SUCCESS [  4.174 s]
[INFO] Helios Client ...................................... FAILURE [ 24.349 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 29.868 s
[INFO] Finished at: 2017-08-18T14:19:54-04:00
[INFO] Final Memory: 41M/661M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.16:test (default-test) on project helios-client: There are test failures.
[ERROR]
[ERROR] Please refer to /Users/mattbrown/code/helios/helios-client/target/surefire-reports for the individual test results.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :helios-client
```

--- a/helios-client/src/main/java/com/spotify/helios/common/Json.java
+++ b/helios-client/src/main/java/com/spotify/helios/common/Json.java
@@ -51,7 +51,10 @@ public class Json {
       .configure(FAIL_ON_UNKNOWN_PROPERTIES, false);
 
   private static final ObjectWriter NORMALIZING_OBJECT_WRITER = new ObjectMapper()
-      .setSerializationInclusion(JsonInclude.Include.NON_EMPTY)
+      .setDefaultPropertyInclusion(
+          //parameters are valueIncl=NON_EMPTY, contentIncl=ALWAYS
+          JsonInclude.Value.construct(JsonInclude.Include.NON_EMPTY, JsonInclude.Include.ALWAYS)
+      )
       .configure(SORT_PROPERTIES_ALPHABETICALLY, true)
       .configure(ORDER_MAP_ENTRIES_BY_KEYS, true)
       .configure(WRITE_DATES_AS_TIMESTAMPS, false)

--- a/pom.xml
+++ b/pom.xml
@@ -157,32 +157,32 @@
             <dependency>
                 <groupId>com.fasterxml.jackson.dataformat</groupId>
                 <artifactId>jackson-dataformat-yaml</artifactId>
-                <version>2.8.8</version>
+                <version>2.9.0</version>
             </dependency>
             <dependency>
                 <groupId>com.fasterxml.jackson.datatype</groupId>
                 <artifactId>jackson-datatype-guava</artifactId>
-                <version>2.8.8</version>
+                <version>2.9.0</version>
             </dependency>
             <dependency>
                 <groupId>com.fasterxml.jackson.core</groupId>
                 <artifactId>jackson-core</artifactId>
-                <version>2.8.8</version>
+                <version>2.9.0</version>
             </dependency>
             <dependency>
                 <groupId>com.fasterxml.jackson.core</groupId>
                 <artifactId>jackson-annotations</artifactId>
-                <version>2.8.8</version>
+                <version>2.9.0</version>
             </dependency>
             <dependency>
                 <groupId>com.fasterxml.jackson.core</groupId>
                 <artifactId>jackson-databind</artifactId>
-                <version>2.8.8</version>
+                <version>2.9.0</version>
             </dependency>
             <dependency>
                 <groupId>com.fasterxml.jackson.jaxrs</groupId>
                 <artifactId>jackson-jaxrs-json-provider</artifactId>
-                <version>2.8.8</version>
+                <version>2.9.0</version>
             </dependency>
             <dependency>
                 <groupId>joda-time</groupId>

